src/goto-symex/goto_symex_state.cpp

```diff
@@ -591,68 +598,144 @@ void goto_symex_statet::rename_address(exprt &expr, const namespacet &ns)
+    const if_exprt &if_expr = to_if_expr(expr);
+    auto renamed_cond = rename<level>(if_expr.cond(), ns);
+    auto renamed_true = rename_address<level>(if_expr.true_case(), ns);
+    auto renamed_false = rename_address<level>(if_expr.false_case(), ns);
+
+    if(
+      renamed_cond != if_expr.cond() || renamed_true.has_value() ||
+      renamed_false.has_value())
     {
-      member_exprt &member_expr=to_member_expr(expr);
+      if_exprt result = if_expr;
 
-      rename_address<level>(member_expr.struct_op(), ns);
+      if(renamed_cond != if_expr.cond())
+        result.cond() = std::move(renamed_cond);
 
-      // type might not have been renamed in case of nesting of
-      // structs and pointers/arrays
-      if(
-        member_expr.struct_op().type().id() != ID_symbol_type &&
-        member_expr.struct_op().type().id() != ID_struct_tag &&
-        member_expr.struct_op().type().id() != ID_union_tag)
+      if(renamed_true.has_value())
       {
-        const struct_union_typet &su_type=
-          to_struct_union_type(member_expr.struct_op().type());
-        const struct_union_typet::componentt &comp=
-          su_type.get_component(member_expr.get_component_name());
-        PRECONDITION(comp.is_not_nil());
-        expr.type()=comp.type();
+        result.true_case() = std::move(*renamed_true);
+        result.type() = result.true_case().type();
       }
-      else
-        rename<level>(expr.type(), irep_idt(), ns);
+
+      if(renamed_false.has_value())
+        result.false_case() = std::move(*renamed_false);
+
+      return std::move(result);
+    }
+    else
+      return {};
+  }
+  else if(expr.id() == ID_member)
+  {
+    const member_exprt &member_expr = to_member_expr(expr);
+
+    auto renamed_struct_op = rename_address<level>(member_expr.struct_op(), ns);
+
+    // type might not have been renamed in case of nesting of
+    // structs and pointers/arrays
+    optionalt<typet> renamed_type;
+    if(
+      renamed_struct_op.has_value() &&
+      renamed_struct_op->type().id() != ID_symbol_type &&
+      renamed_struct_op->type().id() != ID_struct_tag &&
+      renamed_struct_op->type().id() != ID_union_tag)
+    {
+      const struct_union_typet &su_type =
+        to_struct_union_type(renamed_struct_op->type());
+      const struct_union_typet::componentt &comp =
+        su_type.get_component(member_expr.get_component_name());
+      PRECONDITION(comp.is_not_nil());
+      renamed_type = comp.type();
     }
     else
+      renamed_type = rename_type<level>(expr.type(), irep_idt(), ns);
+
+    if(renamed_struct_op.has_value() || renamed_type.has_value())
+    {
+      member_exprt result = member_expr;
+
+      if(renamed_struct_op.has_value())
+        result.struct_op() = std::move(*renamed_struct_op);
+
+      if(renamed_type.has_value())
+        result.type() = std::move(*renamed_type);
+
+      return std::move(result);
+    }
+    else
+      return {};
+  }
+  else
+  {
+    exprt result = expr;
```

Comment:
This copy is sometimes not used, but I don't know if there is a simple way around it.

Reply:
Done.